### DISCUSSION TITLE
Files Drag'n'Drop

### DISCUSSION THREAD

**Discussion by MichaelSvejcar** [AvaloniaUI/Avalonia#10602] [Q&A] [answered]
Hello,

I am trying to implement drag'n'drop logic for one of controls in my Avalonia app. However, no matter how I try, I still get error "Unable to resolve suitable regular or attached property Drop.." Can someone perhaps help me with telling me what am I doing bad? There is not enough documentation for me to figure this out.

This is the xaml code:
`<DataGrid Name="FilesDataGrid" DragDrop.AllowDrop="True" DragDrop.Drop="{Binding FilesDataGrid_Drop}"/>`

And this is in the view model that is set as datacontext:
```
public void FilesDataGrid_Drop(object sender, DragEventArgs e)
        {

        }
```
Binding other methods and events to different controls this way works just fine.

Thank you and wish you all nice day :)

**Comment by timunie** [comment, marked as answer]
I think you need to add this event from code behind like this: 

```cs

    public MyView()
    {
        InitializeComponent();
        
        AddHandler(DragDrop.DragOverEvent, DragOver);
        AddHandler(DragDrop.DropEvent, Drop);
    }
    
    private void Drop(object? sender, DragEventArgs e)
    {
        if (e.Data.GetFileNames() is {} fileNames && ViewModel is not null)
        {
            foreach (var file in fileNames)
            {
                ViewModel.Attachments.Add(new AttachmentLink(file));
            }
        }
    }
```

Something we maybe need to revisit for 11.0 docs.

**Reply by MichaelSvejcar** [reply]
Yes, this is exactly what was needed. Thank you :)

**Comment by MichaelSvejcar** [comment]
@timunie Do you know why DragLeave doesn't work, when I implement it the same way? The event doesn't get fired.

```
public MyView()
    {
        InitializeComponent();
        
        AddHandler(DragDrop.DragOverEvent, DragOver);
        AddHandler(DragDrop.DropEvent, Drop);
        AddHandler(DragDrop.DragLeaveEvent, DragLeave);
    }
```

```
private void DragLeave(object? sender, DragEventArgs e)
        {
            // do logic
        }
```

**Reply by timunie** [reply]
Hm no idea. Can you tell me which version you use? Maybe it's a bug in that version. Also maybe a minimal sample can help to get the root cause for the issue.

**Reply by timunie** [reply]
works for me using 11.0 preview. I made this: 

```cs
    public MainWindow()
    {
        InitializeComponent();
        
        AddHandler(DragDrop.DragOverEvent, DragOver);
        AddHandler(DragDrop.DropEvent, Drop);
        AddHandler(DragDrop.DragLeaveEvent, DragLeave);
    }

    private void DragOver(object? sender, DragEventArgs e)
    {
        Debug.WriteLine("DragOver");
    }

    private void Drop(object? sender, DragEventArgs e)
    {
        Debug.WriteLine("Drop");
    }

    private void DragLeave(object? sender, DragEventArgs e)
    {
        Debug.WriteLine("Drag leave");
    }
```

Debug Output: 
```
DragOver
Drag leave
DragOver
```

**Reply by MichaelSvejcar** [reply]
I am using the latest stable, i.e. 10.18, where it doesn't seem to work currently.

**Reply by Reelix** [reply]
Ran into this identical issue using Avalonia 11.3.0.0 using VSCode 1.100.2 on Linux Mint 22.1

Code compiles without issue, but events don't fire.

**Reply by stevemonaco** [reply]
@Reelix 

There's no native integration for Linux drag-and-drop for files yet, so it's unsupported.

**Reply by Reelix** [reply]
... That would explain it. Thanks. Is there any bug that I can follow that tracks this issue?

**Reply by stevemonaco** [reply]
@Reelix 

There are a few:
https://github.com/AvaloniaUI/Avalonia/issues/6085
https://github.com/AvaloniaUI/Avalonia/issues/14232

**Reply by Reelix** [reply]
Subscribed to those. Much appreciated!
